### ISSUE TITLE
Failure to build on freebsd

### ISSUE THREAD

**Issue by Mythra** [abseil/abseil-cpp#52]
👋 First off, Thanks for this library.

However after setting up a fairly simple test project using Mutex's from abseil I seem to be getting a build failure on my FreeBSD test boxes. Messing around I haven't been able to figure out what's happening yet, but it's also two in the morning on a test project.

The build failure is:

```bash
$ bazel build //span:span --verbose_failures --sandbox_debug
...................
INFO: Analysed target //span:span (20 packages loaded).
INFO: Found 1 target...
ERROR: /usr/home/ec2-user/.cache/bazel/_bazel_ec2-user/3db7863a462abc38a334d1d40287c767/external/com_google_absl/absl/synchronization/BUILD.bazel:47:1: C++ compilation of rule '@com_google_absl//absl/synchronization:synchronization' failed (Exit 1): process-wrapper failed: error executing command 
  (cd /usr/home/ec2-user/.cache/bazel/_bazel_ec2-user/3db7863a462abc38a334d1d40287c767/execroot/vivian && \
  exec env - \
    PWD=/proc/self/cwd \
  /usr/home/ec2-user/.cache/bazel/_bazel_ec2-user/3db7863a462abc38a334d1d40287c767/execroot/vivian/_bin/process-wrapper '--timeout=0' '--kill_delay=15' /usr/local/bin/clang38 -U_FORTIFY_SOURCE '-D_FORTIFY_SOURCE=1' -fstack-protector -Wall -fno-omit-frame-pointer '-std=c++0x' -MD -MF bazel-out/local_freebsd-fastbuild/bin/external/com_google_absl/absl/synchronization/_objs/synchronization/external/com_google_absl/absl/synchronization/blocking_counter.pic.d '-frandom-seed=bazel-out/local_freebsd-fastbuild/bin/external/com_google_absl/absl/synchronization/_objs/synchronization/external/com_google_absl/absl/synchronization/blocking_counter.pic.o' -fPIC -D__CLANG_SUPPORT_DYN_ANNOTATION__ -iquote external/com_google_absl -iquote bazel-out/local_freebsd-fastbuild/genfiles/external/com_google_absl -iquote external/bazel_tools -iquote bazel-out/local_freebsd-fastbuild/genfiles/external/bazel_tools -iquote external/com_googlesource_code_cctz -iquote bazel-out/local_freebsd-fastbuild/genfiles/external/com_googlesource_code_cctz -isystem external/bazel_tools/tools/cpp/gcc3 -isystem external/com_googlesource_code_cctz/include -isystem bazel-out/local_freebsd-fastbuild/genfiles/external/com_googlesource_code_cctz/include -Wall -Wextra -Wcast-qual -Wconversion-null -Wmissing-declarations -Wno-sign-compare -Woverlength-strings -Wpointer-arith -Wunused-local-typedefs -Wunused-result -Wvarargs -Wvla -Wwrite-strings -no-canonical-prefixes -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -c external/com_google_absl/absl/synchronization/blocking_counter.cc -o bazel-out/local_freebsd-fastbuild/bin/external/com_google_absl/absl/synchronization/_objs/synchronization/external/com_google_absl/absl/synchronization/blocking_counter.pic.o)
In file included from external/com_google_absl/absl/synchronization/blocking_counter.cc:15:
In file included from external/com_google_absl/absl/synchronization/blocking_counter.h:24:
In file included from external/com_google_absl/absl/synchronization/mutex.h:79:
external/com_google_absl/absl/synchronization/internal/mutex_nonprod.inc:215:35: error: use of undeclared identifier 'base'
  explicit SynchronizationStorage(base::LinkerInitialized) {}
                                  ^
1 error generated.
Target //span:span failed to build
INFO: Elapsed time: 9.365s, Critical Path: 2.71s
FAILED: Build did NOT complete successfully
```

Clang Version Info:

```
$ /usr/local/bin/clang38 --version
clang version 3.8.1 (tags/RELEASE_381/final)
Target: x86_64-unknown-freebsd11.1
Thread model: posix
InstalledDir: /usr/local/llvm38/bin
```

Bazel WORKSPACE Inclusion of Abseil:
```
http_archive(
    name = "com_google_absl",
    url = "https://github.com/abseil/abseil-cpp/archive/ae0cef3.tar.gz",
    strip_prefix = "abseil-cpp-ae0cef35ae5a7bfd873fb87babdd4900cbdb073c",
    sha256 = "d0182ae6ecd0b88c65b2fe5e1b1fcd42366d9c0f6ac1636cb141016a153b2b35"
)
```

And the actual BUILD:

```
cc_library(
  name = "span",
  srcs = glob([
    "src/**/*.cpp",
  ]),
  hdrs = glob([
    "src/**/*.hh",
  ]),
  deps = [
    "@com_google_absl//absl/synchronization",
    "@glog_repo//:glog",
  ],
  copts = [
    "-std=c++14",
  ],
  strip_include_prefix = "src/",
)
```

**Comment by derekmauro** [comment]
I'm sorry to be the bearer of bad news, but the current policy is that FreeBSD if unsupported and our policy says that patches for unsupported platforms will not be accepted: https://abseil.io/docs/cpp/platforms/platforms

For what it's worth, I didn't make this policy and I don't happen to personally agree with it.  It is also possible that this might change in the future.

Now I don't have a FreeBSD box I can test on, but I bet if you add <code>defined(\_\_FreeBSD\_\_)</code> here <https://github.com/abseil/abseil-cpp/blob/ae0cef35ae5a7bfd873fb87babdd4900cbdb073c/absl/base/config.h#L257> it might work *hint* *hint*.

As a separate issue, base::LinkerInitialized is undefined on systems that don't have mmap.  We should fix this.

**Comment by Mythra** [comment]
@derekmauro I see on that page linux 64 bit is supported? I Ctrl-F'd for "BSD", and didn't see anything on that page. I understand not wanting to support it (just curious what I missed)?

I'll take a look into fixing this issue with the hints you gave me. Thanks for that!

**Comment by Mythra** [comment]
To note on this for others that are curious adding FreeBSD to that config list did fix that part of compiling, however I also had to include: `-lm` to the linkopts to actually allow abseil synchronization (and a few others) to compile too.

I haven't looked into the other compile issues at all yet, but I'm planning on doing that sometime later.

**Comment by derekmauro** [comment]
Good news! We talked about this and decided we can accept PRs for this!

https://github.com/abseil/abseil-cpp/pull/57 is in review.

**Comment by derekmauro** [comment]
I just merged #57. On FreeBSD you should be able to build with `bazel build --linkopt=-lm --linkopt=-pthread //path/to:target` now. We are discussing the best way to get rid of the need to specify --linkopt as part of the build command.
